pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -84,14 +86,83 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
 	return stepResolver
 }
 
+// replacementChainEndsInFailure returns true if the last entry in a chain of operators is in the failed state.
+func replacementChainEndsInFailure(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion) bool {
+	if csv == nil {
+		return false
+	}
+	itr := csv
+	visited := map[string]struct{}{}
+	for {
+		// Visit the csv
+		visited[itr.GetName()] = struct{}{}
+
+		// Check for a replacement
+		next, ok := csvToReplacement[itr.GetName()]
+		if !ok {
+			break
+		}
+
+		// Check if we have visited the CSV before
+		if _, ok := visited[next.GetName()]; ok {
+			break
+		}
+
+		// Move iterator along replacement chain
+		itr = next
+	}
+	return itr.Status.Phase == v1alpha1.CSVPhaseFailed
+}
+
+func (r *OperatorStepResolver) CachePredicates(namespace string) ([]cache.Predicate, error) {
+	predicates := []cache.Predicate{}
+
+	operatorGroupList, err := r.client.OperatorsV1().OperatorGroups(namespace).List(context.TODO(), metav1.ListOptions{})
```

Comment:
Looking at all of these OG's query in here and other places, it makes me wonder if `UpgradeStrategy` should be in Subscription instead of OG given resolver is working with Subscription so we don't need to query OG at all. Too late for it now so it's just me stating my thought here.

Reply:
May as well add a OGlister to OperatorStepResolver for the sake of efficiency.

Reply:
Refactored to remove the need to list existing operatorGroups.